File: src/domain/feeKey.ts
import type { AssetClass, Channel, FeeKey, Product, Session } from './types';

export function isDerivative(a: AssetClass): boolean { return a === '국내파생' || a === '해외파생'; }

// feeKey를 차원 값에서 직접 조립한다. 주식형(파생 아님)은 품목 차원이 붕괴되므로 productCode를
// 넘겨도 product는 null이 된다 — 화면이 종목을 고르지 않고 거래소·세션·채널만으로 해석할 수 있게.
export function buildFeeKey(
  assetClass: AssetClass, exchange: string, session: Session, channel: Channel, productCode?: string | null,
): FeeKey {
  return {
    assetClass, exchange, session, channel,
    product: isDerivative(assetClass) ? (productCode ?? null) : null,
  };
}

export function deriveFeeKey(product: Product, session: Session, channel: Channel): FeeKey {
  return buildFeeKey(product.assetClass, product.exchange, session, channel, product.code);
}

export function feeKeyString(k: FeeKey): string {
  const base = `${k.assetClass}|${k.exchange}|${k.session}|${k.channel}`;
  return k.product ? `${base}|${k.product}` : base;
}

File: server/src/main/java/kr/fees/domain/FeeKey.java
package kr.fees.domain;

/**
 * 조회키 — 배정판/랭킹이 다루는 셀 좌표. 파생만 product 차원을 가지며,
 * 주식형은 of() 팩토리에서 product 를 null 로 강제한다(v0.5 불변식: 주식은 종목 차원 없음).
 * exchange/session/channel 은 '*'(전체) 허용, product 는 파생일 때 품목코드 또는 '*'.
 */
public record FeeKey(
    AssetClass assetClass,
    String exchange,
    LookupKey lookupKey,
    String session,
    String channel,
    String product
) {
    public static FeeKey of(AssetClass assetClass, String exchange, LookupKey lookupKey,
                            String session, String channel, String productCode) {
        String product = assetClass.isDerivative() ? productCode : null;
        return new FeeKey(assetClass, exchange, lookupKey, session, channel, product);
    }
}
